pkg/controller/podautoscaler/horizontal.go

```diff
@@ -538,19 +535,40 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+// stabilizeRecommendation:
+// - records the newest recommendation
+// - removes recommendations older than ???
+// - returns max of the remaining recommendations
+func (a *HorizontalController) stabilizeRecommendation(key string, prenormalizedDesiredReplicas int32) int32 {
+	a.recommendations[key] = append(a.recommendations[key], timestampedRecommendation{prenormalizedDesiredReplicas, time.Now()})
```

Comment:
should we just have a basic circular buffer built on a slice here?  it's probably not a big deal, but a circular buffer seems like it would be much more space/garbage efficient, and not particularly more complex...

Reply:
From user point of view what matters is for how much time we stabilize size. So I like this approach because it directly checks the thing user cares about.

I don't want to use circular buffer. I can think of two ways it may start behaving badly:
- HPA starts running less frequently (for example it's busy, can't finish on time).
- HPA starts running more frequently (someone tweaks --horizontal-pod-autoscaler-sync-period).

There are solutions to those but I'm not confident I thought of all the potential problems. And I don't want to mess up operation of HPA because I missed some corner case while doing minor performance tweak.

BTW I checked with simple loop:

```
slice := []int{1,2,3,4,5}
for {
  slice = append(slice, rand.Int())
		slice = slice[1:]
}
```

that this does not leek (after getting through 9.7e9 iterations of the loop process was still using ~10kB memory).

Reply:
> that this does not leek (after getting through 9.7e9 iterations of the loop process was still using ~10kB memory).

testing locally on my system, I see it creeping quite a bit higher than that before leveling off, which makes me a bit uncomfortable, especially since a circular buffer is `sizeof(int) * (1 + numSamplesToSave)` give-or-take.  Theoretically, it shouldn't *leak*, since each iteration is equivalent to

```
if len(slice) + 1 > cap(slice) {
  newSlice := make([]int, cleverResizeLengthThatsAtLeastLargeEnough)
  copy(newSlice, oldSlice)
  forgetAbout(oldSlice)
  newSlice.start = 1
}
```

So it'll end up resizing cleverly once, and then just creating slices of that size forever, but we'll end up creating a lot of garbage.  It's probably not a huge deal, but it does leave me a little nervous about what happens if we have 10k HPAs.

> HPA starts running more frequently

This can't happen in-place, so it shouldn't be problematic

> HPA starts running less frequently

This also shouldn't be terribly problematic -- we could still keep around times, and just double-check to age out super-old points.

Reply:
Yes, the problems I spotted are easy to fix. My point is that I'm worried that I missed some other problems or that someone in the future will miss a problem and as a result HPA will not stabilize scale downs properly.

Reply:
@DirectXMan12 please share numbers you got when you checked my code for leaking. They'll be useful for considering trade off between memory usage and potential of algorithm malfunction.

Reply:
Looking at the Go Heap allocation stats (runtime.ReadMemStats), I was seeing a baseline of ~115KiB, and then seeing it go up to 4-8MiB before leveling off, compared to using a basic circular buffer with a slice, where it stayed around ~115KiB.

Not a huge deal if that's the end of the story, but it does make me a bit nervous that it could end up causing some weird behavior in a real-world scenario. /me shrugs

Reply:
Ok, I think I have a solution that has all the good traits of our solutions:

```
// Working with historical recommenedations stored in
// recommendations []timestampedSample
// and
// latestRecommendation timestampedSample

foundOldSample := false
oldSampleIndex := 0
stabilizedRecommendation := latestRecommendation.value
for i, sample := range recommendations {
  if sample.timestamp.Before(cutoff) {
    foundOldSample = true
    oldSampleIndex = i
  } else {
    if sample.value > stabilizedRecommendation {
      stabilizedRecommendation = sample.value
    }
  }
}
if foundOldSample {
  recommendations[oldSampleIndex] = latestRecommendation
} else {
  recommendations = append(recommendations, latestSample)
}
```

This:
- during normal operation keeps using the same slice , without resizing it
- increases size of the slice if it needs to  do it to avoid discarding relevant historical information

Reply:
That works, I think.  With your scheme, we'll get out-of-order timestamps though.  Might just want to keep around a "nextWritable" index, and resize if the nextWritable is still within the window that we care about.

Reply:
I don't mind that (though I don't see much benefit in keeping samples in order either).

(mentioning @krzysztof-jastrzebski for visibility)

Reply:
yeah, it doesn't matter much for the way you wrote it.  I'm fine with not keeping them in order as well.

Reply:
@krzysztof-jastrzebski up is pseudo code that looks good to me & @DirectXMan12 as well as a suggestion on how to keep recommendations in order if you think it's worth the extra complexity.